packages/components/src/data-model/word-query/word-query.js

```diff
@@ -0,0 +1,314 @@
+import WordQueryError from '@comp/data-model/word-query/word-query-error.js'
+
+/**
+ * Contains all the business logic that is necessary to resolve the GraphQL word query.
+ */
+export default class WordQuery {
+  constructor (options = {}, queryParams = {}) {
+    this._context = {
+      queryParams,
+      options: {
+        language: Language.fromJsonObject({ code: queryParams.language }),
+        clientID: options.clientID,
+        lexicons: options.lexicons,
+        shortLexicons: options.shortLexicons
+      },
+      cancelled: false,
+      result: undefined
+    }
+  }
+
+  start (result) {
+    this._context.result = result
+    this._context.result.loadingStarted()
+    if (this._context.queryParams.getLexemes) { this._context.result.state.lexemes.loading = true }
+    if (this._context.queryParams.getShortDefs) { this._context.result.state.shortDefs.loading = true }
+    const executionPath = WordQuery._createExecutionPath(this._context.queryParams)
+    try {
+      // Ignore result of an asynchronous operation and return immediately
+      // Result will be tracked via the proxied `result` variable
+      this._context = WordQuery._execute(executionPath, this._context)
+    } catch (err) {
+      Logger.getInstance().error(err)
+    }
+    return result
+  }
+
+  cancel () {
+    this._context.cancelled = true
+  }
+
+  // Creates a set of methods to be executed in order to fulfil the request.
+  static _createExecutionPath (queryParams) {
+    /*
+     Execution path consists of several batches. Batches are execution sequentially: each next batch
+     is executed after the previous one is completed.
+     Each batch consist of one or several jobs. Jobs within a batch can be executed in parallel.
+     */
+    let path = [] // eslint-disable-line prefer-const
+
+    let getLexemesBatch = [] // eslint-disable-line prefer-const
+    if (queryParams.getLexemes) {
+      if (queryParams.useMorphService) {
+        getLexemesBatch.push('_getTuftsMorphology')
+      }
+      if (queryParams.useTreebankData) {
+        getLexemesBatch.push('_getTreebankData')
+      }
+      if (queryParams.useWordAsLexeme) {
+        getLexemesBatch.push('_getWordAsLexeme')
+      }
+    }
+    if (getLexemesBatch.length > 0) {
+      path.push(getLexemesBatch)
+    }
+    path.push(['_disambiguate'])
+
+    if (queryParams.getShortDefs) {
+      path.push(['_getShortDefs'])
+    }
+    return path
+  }
+
+  // Runs all the jobs within the execution path.
+  static async _execute (executionPath, context) {
+    try {
+      for (const batch of executionPath) {
+        if (!context.cancelled) {
+          const jobs = batch.map(job => WordQuery[job](context))
+          await Promise.all(jobs)
+        }
+      }
+    } catch (err) {
+      Logger.getInstance().error(err)
+    }
+    WordQueryResult.loadingStopped(context.result.state)
+    context.result.state = Object.assign({}, context.result.state)
```

Comment:
Yes, exactly. The proxy attached to the result object (`context.result`) tracks only the top-level props of the object: `context.result.homonymGroup`, `context.result.state`, and `context.result.errors`. If we assign to something below that level, as `context.result.state.loading = false`, the change will be not noticed.

So the key here is not to create a copy of an object, but rather to assign to the top-level prop of the `context.result`. The assignment as the one referenced above is the way I came up with that makes sense from code point of view. The other alternative that, I think, I've used somewhere else is a self-assignment: `context.result.state = context.result.state`. This works (the change is tracked) but it's a practice that can be frowned upon and not understood easily.

Actually, I think there is a problem with this code above because `Object.assign()` does a shallow copy only, and here we need a deep one. So I think I need to fix that.

Another way to create a copy is using `JSON.parse(JSON.stringify(context.result.state))`. But that is slower and looks probably more obscure.

So what would be the best way to do an assignment to `context.result.state`? Use `context.result.state = context.result.state` and explain in the comment why is it so? Or use a JSON method? Are there any other solutions? What do you think?

Reply:
what do you think about adding a method that does just the same thing, but is named accordingly and then call that whenever we need to make a deep copy for reactivity? that would make the code clearer.

E.g. `WordQueryRequest.ensureReactivity(object)`

Reply:
That's a great idea! Will do that